### PR TITLE
Tighten assert in reverse

### PR DESCRIPTION
**Pull request by dsyme**
Address issues raised about #235 in review #258

* add test that shows slicing fix in #235 fixes a real problem (I've verified TestDerivativeSliceTFinalColumn failed before #235)

* add test that shows that `boundsToShape` change in #235 was not necessary (it was actually a no-op brought over mistakenly from #237), and revert that

* show we can tighten the assert added in #235 per discussion in #258 to prove there's no strange broadcasting going on.

### CODE REVIEW

File: tests/DiffSharp.Tests/TestDerivatives.fs [hunk 1]
```diff
@@ -2354,6 +2354,20 @@ type TestDerivatives () =
             Assert.True(revz.allclose(revzCorrect, 0.01))
             Assert.True(revxd.allclose(revxdCorrect, 0.01))
 
+
+    [<Test>]
+    member _.TestDerivativeSliceTFinalColumn () =
+        for combo in Combos.AllDevicesAndBackendsFloat32 do
+            let revx = combo.tensor([54.7919; 70.6440; 16.0868]).reverseDiff()
```
Comment (gbaydin) [inline tests/DiffSharp.Tests/TestDerivatives.fs:2361]: We need the forward derivative test

File: tests/DiffSharp.Tests/TestDerivatives.fs [hunk 2]
```diff
@@ -2409,6 +2423,21 @@ type TestDerivatives () =
             Assert.True(revxd.allclose(revxdCorrect, 0.01))
             Assert.True(revyd.allclose(revydCorrect, 0.01))
 
+    [<Test>]
+    member _.TestDerivativeSliceT_2D () =
+        for combo in Combos.AllDevicesAndBackendsFloat32 do
```
Comment (gbaydin) [inline tests/DiffSharp.Tests/TestDerivatives.fs:2428]: We need the forward derivative test

File: tests/DiffSharp.Tests/TestDerivatives.fs (hunk 1 above)
Comment (gbaydin) [reply tests/DiffSharp.Tests/TestDerivatives.fs:2361]: We also need at least one test for >1 dimensional case

**Comment by dsyme** [comment]
Cool I'll add systematic slicing testing tomorrow

**Comment by gbaydin** [comment]
> Cool I'll add systematic slicing testing tomorrow

I mean I think just a 2d case would suffice to convince us that the implementation generalizes to things > 1d. I didn't want to propose spending too much time with this.

**Pull request merged**